#!/usr/bin/env python
# -*- coding:UTF-8 -*-
__author__ = 'Zhengtao Xiao'

import argparse
import os

def parsing_transcript():
	parser = argparse.ArgumentParser(
		description="This script is designed for preparing transcripts annotation files."
	)
	parser.add_argument("-g","--gtf",dest="gtfFile",required=True,type=str,
	                    help='Default, suitable for GENCODE and ENSEMBL GTF file, \
	                          please refer: https://en.wikipedia.org/wiki/GENCODE')
	parser.add_argument("-f","--fasta",dest="genomeFasta",required=True,type=str,
	                    help="The genome sequences file in fasta format.")
	parser.add_argument("-o","--out_dir",required=True,type=str,dest="out_dir",help="annotation directory name.")
	args = parser.parse_args()

	if not os.path.exists(args.out_dir):
		try:
			os.mkdir(args.out_dir)
		except OSError,e:
			raise OSError(e)

	if not os.path.exists(args.gtfFile):
		raise ValueError("Error, gtf file not found:%s.\n" % args.gtfFile)

	if not os.path.exists(args.genomeFasta):
		raise ValueError("Error, genomic fasta not found: %s\n" % args.genomeFata)

	return args
def parsing_metaplots():
	parser = argparse.ArgumentParser(
		description="""
		This script create aggregate plots of distances from the 5'end of reads to start or stop codons,
		which help determine the length range of the PRF reads that are most likely originated from the
		translating ribosomes and identify the P-site locations for each reads lengths.
		"""
	)
	parser.add_argument("-a","--annot_dir",dest="annot_dir",required=True,type=str,
	                    help="transcripts annotation directory, generated by prepare_transcripts.")
	parser.add_argument("-r","--ribo_bam",dest="ribo_bam",required=True,type=str,
	                    help="ribo-seq alignment bam/sam file at the transcript-level.")
	parser.add_argument("-s","--stranded",dest="stranded",required=False,type=str,choices=["yes","reverse"],
	                    default="yes",help="whether the data is strand-specific, \
	                    reverse means reversed strand interpretation.")
	parser.add_argument("-m","--minimum-length",dest="minLength",required=False,type=int,default=24,
	                    help="minimum read length for analysis, default 24")
	parser.add_argument("-M","--maximum-length",dest="maxLength",required=False,type=int,default=35,
	                    help="maximum read length for analysis, default 35")
	parser.add_argument("-o","--outname",dest="outname",required=False,type=str,default="metaplots",
	                    help="name of output pdf file(default: metaplots)")
	args = parser.parse_args()

	if not os.path.exists(args.annot_dir):
		raise ValueError("Error, the transcript annotation directory not found: {} \n \
		                  pleas run prepare_transcripts.py first.".format(args.annot_dir))
	if args.minLength > args.maxLength:
		raise ValueError("minimum length must be <= maximum length (currently %d and %d, respectively)" % (args.minLength, args.maxLength))
	if args.minLength <= 0 or  args.maxLength <=0:
		raise ValueError("minimum length or maximum length must be larger than 0.")
	if not os.path.exists(args.ribo_bam):
		raise  ValueError("Error, the bam file not found: %s\n" % args.ribo_bam)
	args.stranded = True if args.stranded == "yes" else False
	return args

def parsing_ribo():
	parser = argparse.ArgumentParser(
		description="This program is designed for detecting ORF using ribosome-profiling data."
	)
	parser.add_argument("-a","--annot_dir",dest="annot_dir",required=True,type=str,
	                    help="transcripts annotation directory, generated by prepare_transcripts.")
	parser.add_argument("-c","--config_file",dest="config_file",required=True,type=str,
	                    help="list bam file and P-sites information in this file, \
	                    please refer to the provided sample file in data folder.")
	# parser.add_argument("-n","--num-threads",dest="threads_num",default=1,required=False,
	#                     help="number of threads, optimal number is number of bam files.", type=int)
	parser.add_argument("-l","--longest-orf",dest="longest_orf",choices=["yes","no"],default="yes",required=False,
	                    help="Default: yes, the region from most distal AUG to stop was defined as an ORF. \
	                          no, judge by program.", type=str)
	parser.add_argument("-p","--pval-cutoff",dest="pval_cutoff",default=0.01,required=False,
	                    help="P-value cutoff for ORF filtering.", type=float)
	parser.add_argument("-s","--start_codon",default="ATG",type=str,dest="start_codon",
	                    help="The canonical start codon. default: ATG")
	parser.add_argument("-A","--alt_start_codons",default="",type=str,dest="alternative_start_codons",
	                    help="The codons considered as possible translation initiation sites. (Default: None). List more separated by comma.")
	parser.add_argument("-S","--stop_codon",default="TAA,TAG,TGA",type=str,dest="stop_codon",
	                    help="Stop codon, default: TAA,TAG,TGA.")
	parser.add_argument("-t","--transl_table",default=1,dest="transl_table",type=int,
	                    help="ORF translation table(Default: 1). Assign the correct genetic code based on your organism, \
	                    [please refer: https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi]")
	parser.add_argument("-m","--min-AA-length",dest="min_AA_length",default="20",required=False,
	                    help="The minimal length of predicted peptides", type=int)
	parser.add_argument("-o","--output-file",dest="output_file",default="final_result",required=False,
	                    help="output file name, default: final_result", type=str)

	args = parser.parse_args()

	if not os.path.exists(args.annot_dir):
		raise  ValueError("Error, the transcript annotation directory not found: {} \n \
		             pls run prepare_transcripts.py first. ".format(args.annot_dir))

	return args

def plot_orf_density():
	parser = argparse.ArgumentParser(
		description="This script is designed for plot the P-site profile of specified ORF."
	)
	parser.add_argument("-a","--annot_dir",dest="annot_dir",required=True,type=str,
	                    help="transcripts annotation directory, generated by prepare_transcripts.")
	parser.add_argument("-c","--config_file",dest="config_file",required=True,
	                    help="defile bam file information in this file, \
	                    please refer to the provided sample file in data folder.",type=str)
	parser.add_argument("-t","--transcript_id",dest="transcript_id",required=True,type=str,
	                    help="the transcript id")
	parser.add_argument("-s","--orf_tstart",dest="orf_tstart",required=True,type=int,
	                    help="the start position in transcript(orf_tstart)")
	parser.add_argument("-e","--orf_tstop",dest="orf_tstop",required=True,type=int,
	                    help="the stop position in transcript(orf_tstop)")
	parser.add_argument("-o","--outname",dest="outname",required=False,type=str,default="",
	                    help="output file name,default is transcriptid_tstart_tstop.pdf")
	args = parser.parse_args()

	args.orf_tstart = args.orf_tstart -1 # change to 0 based
	if not os.path.exists(args.annot_dir):
		raise  ValueError("Error, the transcript annotation directory not found: {} \n \
		             pls run prepare_transcripts.py first. ".format(args.annot_dir))
	return args
